retriever/lib/engine_tools.py

```diff
@@ -101,37 +103,74 @@ def reset_retriever(scope="all", ask_permission=True):
             print("can't find script {scp}".format(scp=scope))
 
 
-def json2csv(input_file, output_file=None, header_values=None, encoding=ENCODING):
-    """Convert Json file to CSV.
-
-    Function is used for only testing and can handle the file of the size.
-    """
+def json2csv(input_file, output_file=None, header_values=None, encoding=ENCODING, row_key=None):
+    """Convert Json file to CSV."""
     file_out = open_fr(input_file, encoding=encoding)
     # set output file name and write header
     if output_file is None:
         output_file = os.path.splitext(os.path.basename(input_file))[0] + ".csv"
     csv_out = open_fw(output_file, encoding=encoding)
     if os.name == 'nt':
-        outfile = csv.DictWriter(csv_out,
+        outfile = csv.writer(csv_out,
                                  dialect='excel',
                                  escapechar="\\",
                                  lineterminator='\n',
-                                 fieldnames=header_values)
+                                )
     else:
-        outfile = csv.DictWriter(csv_out,
+        outfile = csv.writer(csv_out,
                                  dialect='excel',
                                  escapechar="\\",
-                                 fieldnames=header_values)
+                                )
     raw_data = json.loads(file_out.read())
-    outfile.writeheader()
-
-    for item in raw_data:
-        outfile.writerow(item)
+    raw_data = walker(raw_data, row_key=row_key, header_values=header_values, rows=[], normalize=False)
+    raw_data = [row.tolist() for row in raw_data]
+    outfile.writerow(header_values)
+    outfile.writerows(raw_data)
     file_out.close()
     subprocess.call(['rm', '-r', input_file])
     return output_file
 
 
+def walker(dictionary, row_key=None, header_values=None, rows=[], normalize=False):
+    """
+    Both name and columns cant be none
+    """
+    #  Handles the simple case, where row_key and column_key are not required
+    if not row_key and not header_values:
+        if isinstance(dictionary, dict):
+            # rows = pd.Series(dictionary)
+            rows = pd.DataFrame([dictionary]).values
+            return rows
+        elif isinstance(dictionary, list):
+            rows = pd.DataFrame(dictionary, columns=header_values).values
+            return rows
+
+    if isinstance(dictionary, dict):
+        if header_values and (header_values in list(dictionary.keys()) or header_values == list(dictionary.keys())):
```

Comment:
Use `header_values in dictionary` instead of `header_values in list(dictionary.keys())`. Hash lookup is faster.

Reply:
header_values is a `list` that is why `header_values in dictionary` will not work.

Reply:
header_values is a `list` that is why `header_values in dictionary` will not work.

Reply:
I may be wrong but if it's a list looking at the code it looks like you are checking if a list exists as one of the dictionary keys. I think even `header_values in list(dictionary.keys())`  will not work then. Can you add a comment showing the values of `header_values` and `dictionary.keys()` here.

Reply:
`['id', 'firstname', 'surname', 'motivation', 'share'] dict_keys(['id', 'firstname', 'surname', 'motivation', 'share'])`
It seems that you are right, the logic I have used here is wrong. I have updated the code to use `sets` instead.